documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,179 @@
+
+For Example:
+
+Our `<prefix>`s would be derived from the environment and are described via IaC
+in the devops repo
+```
+skate/prod
+skate/dev
+skate/dev-blue
+skate/dev-green
+```
+And the final combined topic hierarchy would be formed by prepending _one_ of 
+the `<prefix>`s
+```
+<prefix>/trip_modifications/
+```
+For instance,
+```
+skate/dev-blue/trip_modifications/
+```
+
+
+For each new and unique trip modification, we'll generate a unique ID and use
+that as our Trip Modification identifier, named `id`.
+
+> [!NOTE]  
+> The GTFS-RT Trip Modifications spec does not currently have a ID for each modification,
+> so this `ID` _will not_ appear with the Trip Modification Message.
+
+The topic hierarchy is then formed using this ID as the topic level
+```
+<prefix>/trip_modifications/#{id}/
+```
+
+Then, to create or update an associated `trip_modification`, we'll use the topic
+level `trip_modification`
+Creating the topic
+```
+<prefix>/trip_modifications/#{id}/trip_modification
+```
+
+Any associated [`Shape`](https://gtfs.org/realtime/reference/#message-shape)
+Messages will be published to the topic level `shape` in the hierarchy.
+```
+<prefix>/trip_modifications/#{id}/shape
+```
+
+On `Shape` Messages, the
+[`Shape.shape_id`](https://gtfs.org/realtime/reference/#message-shape)
+field will also have a unique ID generated by Skate, which will be
+referenced by the corresponding Trip Modification's
+[`SelectedTrips.shape_id`](https://gtfs.org/realtime/reference/#message-selectedtrips)
+field.
+
+#### Active Trip Modifications State
+Messages pushed to these topics will have the
+[`retain` property](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/)
+set so that new connections get active Trip Modifications and Shapes published
+to them on connect.
+
+Deactivating a Trip Modification will be done by
+[publishing a zero-byte payload](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/#heading-how-to-delete-retained-messages-in-mqtt)
+each topic within the `<prefix>/trip_modifications/#{id}/`
+topic hierarchy, which will clear any retained messages and ensure that new
+subscribers are not notified of these now deleted messages and existing
+subscribers are made aware of the removal.
+
+### Schema
+We'll publish using JSON encoded versions of the GTFS-RT Messages to these topics.
+The `payload` of the MQTT message will contain a JSONAPI compatible string
+where the `data` key contains the GTFS-RT Message encoded as JSON.
+
+`<prefix>/trip_modifications/#{id}/trip_modification`
+```json5
+{
+  "data": {
+    # <GTFS-RT Trip Modification Message as JSON>
+  }
+  "meta": {
+    # this field will be removed once we give users the ability to activate and
+    # deactivate detours themselves.
+    #
+    # Currently, this indicates that this message should not be published to
+    # applications outside of the MBTA, and is for internal testing and
+    # development only.
```

Comment:
```suggestion
    // this field will be removed once we give users the ability to activate and
    // deactivate detours themselves.
    //
    // Currently, this indicates that this message should not be published to
    // applications outside of the MBTA, and is for internal testing and
    // development only.
```

This is rendering funny, and I think it's because `#` isn't the right comment indicator for JSON.


```json5
{
  # This is not a comment
  "elixir comment": "#",

  // This is a comment
  "json comment": "//",
}
```

Reply:
Ah yep, great catch!! I knew this looked a bit weird, I've gotten to used to not having proper highlighting 😅 

[Thanks a bunch!](https://github.com/mbta/skate/compare/8d768330b70c9ff37649a6a1c99793fc6a59d1ad..dc0d1bb510f12ff35ae2a1ec3940cf1b36c269a5)